### DISCUSSION TITLE
Executing zkAsm code from Rust

### DISCUSSION THREAD

**Discussion by aborg-dev** [0xPolygonHermez/pilcom#66] [Q&A] [answered]
# Context
As a part of ZK WASM compiler (https://github.com/near/wasmtime) written in Rust we are compiling WASM code (tests like [this one](https://github.com/near/wasmtime/blob/main/cranelift/zkasm_data/add.wat)) to zkAsm and want to execute the resulting zkAsm code.

At the moment, we use a nodejs interpreter to run each zkAsm file using [this script](https://github.com/near/wasmtime/blob/main/tests/zkasm/run-tests-zkasm.js). This provides the necessary functionality, but the startup time of `npm run` is around 700ms which would likely be too slow when we have more than 100 tests. We want to find a way to run these tests faster. We can't easily batch the execution into a single invocation of `npm run` as zkAsm files are generated on the fly by Rust tests that run in parallel.

One of the ideas was to use https://deno.com/ runtime to run the JS code. This way it could be possible to [embed](https://crates.io/crates/deno_core) Deno runtime into Rust binary which would remove the overheads of starting a subprocess. I did a [quick prototype](https://github.com/near/wasmtime/pull/210) but it would require some changes to zkAsm codebase.

See https://near.zulipchat.com/#narrow/stream/295306-pagoda.2Fcontract-runtime/topic/Running.20zkAsm.20code.20from.20Rust for full discussion.

# Questions

1. Is there a way to run zkAsm code in Rust? Is there one planned in the future?
2. Have you investigated using Deno to run zkAsm tooling and would that be a good direction to research?

**Comment by Nikopiter** [comment]
Hello and good question!

**Comment by krlosMata** [comment, marked as answer]
I will explain how we run `zkasm` test in parallel.
We use the well know framework `mocha` to run the tests. This framework allows you to run test in parallel by just adding `--parallel` and you can define `--jobs $NUM_CPUS` to use.

There must be a startup phase where all `*.js` mocha files are generated:
- https://github.com/0xPolygonHermez/zkevm-rom/blob/main/package.json#L24
- https://github.com/0xPolygonHermez/zkevm-rom/blob/main/tools/parallel-testing/gen-parallel-tests.js
Then, we just run all of them in parallel:
- https://github.com/0xPolygonHermez/zkevm-rom/blob/main/package.json#L25

The startup phase should be very fast. The testing phase depend pretty much on the test you are running and the machine CPUs
